policy-correlation: enhancements and fixes to correlated policies

sypakine opened the pull request:
Several enhancements to policy correlation:

- add API kind to policy struct
- support cluster-scoped APIs
- Fix

Policy can be derived from multiple policy APIs (k8s np, cnp, ccnp, etc.). This change includes that Kind in the correlated policy.

Including the Kind in the correlated policy simplifies identifying the resource responsible for enforcement.

Release note:

```release-note
Policy correlation:
- include policy API Kind in the policy correlation result
- correlate cluster-scoped APIs (e.g. CCNP)
- correctly correlate L4 and protocol-only policies
```

gandro commented on api/v1/flow/flow.proto:
```diff
@@ -464,6 +464,7 @@ message Policy {
 	string namespace = 2;
 	repeated string labels = 3;
 	uint64 revision = 4;
+    string kind = 5;
```
This seems to be using spaces for indentation, where was the rest of the file uses tabs

sypakine replied:
Ah, thanks. I'll push that change when I resolve the other comment.

gandro commented on pkg/policy/correlation/correlation.go:
```diff
@@ -134,7 +134,31 @@ func lookupPolicyForKey(ep getters.EndpointInfo, key policy.Key, matchType uint3
+		}
+
+		// Check for allow-specific-port-protocol policies.
+		// This covers the case where one or more identities are allowed by network policy.
```
Good catch. However, wouldn't this be the case if `matchType == monitorAPI.PolicyMatchL4Only`?

In other words, shouldn't this be it's own case under `monitorAPI.PolicyMatchL4Only`

sypakine replied:
Thanks for the review. Makes sense to me Let me test that change and get back or upload the changes.

gandro reviewed (commented):
Thank you (and apologies for the late review, I was on vacation)! Looks good overall, two minor things that I think we should address however.

sypakine committed:
feat(policy-correlation): include policy API Kind in the correlated policy

Policy can be derived from multiple policy APIs (k8s np, cnp, ccnp, etc.). This change includes that Kind in the correlated policy.

Including the Kind in the correlated policy simplifies identifying the resource responsible for enforcement.

Signed-off-by: Mark St John <markstjohn@google.com>

sypakine committed:
fix(policy-correlation): lookup L4 (proto+port) policy

Policy rules that only define protocol and port are not fetched during policy correlation. Add a lookup for these types of rules.

Partial Fix for #33919

Signed-off-by: Mark St John <markstjohn@google.com>

sypakine committed:
fix(policy-correlation): correlate cluster-scoped policy APIs.

Update correlation logic to make flow policy fields optional.
Therefore, namespace-less resources (i.e. cluster-scoped resources) will populate the name and kind.

Primarly, this change results in correlating policies enforced via CCNP resources.

Signed-off-by: Mark St John <markstjohn@google.com>

sypakine committed:
fix(policy-correlation): lookup protocol-only policy

Policy rules that only define protocol are not fetched during policy correlation. Add a lookup for these types of rules.

Fix: #33919

Signed-off-by: Mark St John <markstjohn@google.com>

sypakine committed:
test: add test case for match all policy

Increase code coverage by extending unit tests to all (supported) policy match types.

Also, expand the comments to harmonize the description.

Signed-off-by: Mark St John <markstjohn@google.com>

gandro reviewed (approved):
This is awesome and the improved testing coverage is more than welcome! Thank you so much
